src/espidf.rs

```diff
@@ -122,32 +122,58 @@ pub enum FromEnvError {
 /// Information about a esp-idf source and tools installation.
 #[derive(Debug)]
 pub struct EspIdf {
-    /// The esp-idf repository.
-    pub repository: git::Repository,
+    /// The esp-idf source tree.
+    pub tree: SourceTree,
     /// The binary paths of all tools concatenated with the system `PATH` env variable.
     pub exported_path: OsString,
     /// The path to the python executable to be used by the esp-idf.
     pub venv_python: PathBuf,
     /// The version of the esp-idf or [`Err`] if it could not be detected.
     pub version: Result<EspIdfVersion>,
-    /// Whether [`EspIdf::repository`] is installed and managed by [`Installer`] and
-    /// **not** provided by the user.
+    /// Whether [`EspIdf::tree`] is a [`SourceTree::Git`] repository installed
+    /// and managed by [`Installer`] and **not** provided by the user.
     pub is_managed_espidf: bool,
 }
 
+#[derive(Debug, Clone)]
+pub enum SourceTree {
+    Git(git::Repository),
+    Plain(PathBuf),
+}
+
+impl SourceTree {
+    pub fn open(path: &Path) -> Self {
+        git::Repository::open(path)
+            .map(SourceTree::Git)
+            .unwrap_or_else(|_| SourceTree::Plain(path.to_owned()))
+    }
+
+    pub fn path(&self) -> &Path {
+        match self {
+            SourceTree::Git(repo) => repo.worktree(),
+            SourceTree::Plain(path) => path,
+        }
+    }
+}
+
 impl EspIdf {
     /// Try to detect an activated esp-idf environment.
     pub fn try_from_env() -> Result<EspIdf, FromEnvError> {
         // detect repo from $IDF_PATH
         let idf_path = env::var_os(IDF_PATH_VAR).ok_or_else(|| {
             FromEnvError::NoRepo(anyhow!("environment variable `{IDF_PATH_VAR}` not found"))
         })?;
-        let repo = git::Repository::open(idf_path).map_err(FromEnvError::NoRepo)?;
+
+        let idf_absolute_path = crate::cargo::workspace_dir()
+            .map(|workspace| workspace.join(&idf_path))
+            .unwrap_or(PathBuf::from(idf_path));
```

Comment:
Instead of modifying `idf_path` here, lets take it as an `idf_path: Option<&Path>` argument, and modify it in esp-idf-sys. Then, we either use the argument as the idf_path or fall back to the environment variable. This allows us to also support the use-case where `idf_path` is specified in Cargo.toml metadata and is not a git repo.

Additionally, making it an absolute path should only happen if `idf_path` is actually relative (which is probably rarely the case).

Reply:
I've added the `idf_path: Option<&Path>` argument and made the corresponding changes over in https://github.com/haileys/esp-idf-sys/commit/c4b8098b5201a8d7b3ac508edbf1086778f7cc56.

`join` does replace the whole path if its argument is an absolute path, but just for clarity I have used `abspath_relative_to` instead. This is needed for the submodule use case, otherwise the path seems to be treated as relative to esp-idf-sys, not the originating workspace:

```
  Using custom user-supplied esp-idf repository at 'vendor/esp-idf' (detected from env variable `IDF_PATH`)
  python3: can't open file '/home/hailey/code/esp-idf-sys/vendor/esp-idf/tools/idf_tools.py': [Errno 2] No such file or directory
```

Reply:
> `join` does replace the whole path if its argument is an absolute path,

Ohh, good to know. 

> This is needed for the submodule use case, otherwise the path seems to be treated as relative to esp-idf-sys, not the originating workspace:

Yup, still a good change, but embuild shouldn't be concerned with this.